File: src/backend/ai/chatbot.js
// backend/ai/chatbot.js

const Groq = require('groq-sdk');
const { buildSystemPrompt } = require('./prompts/systemPrompt');

const client = new Groq({ apiKey: process.env.GROQ_API_KEY });

async function streamChatResponse(userMessage, userData, res) {
    try {
        const systemPrompt = buildSystemPrompt(userData);

        res.setHeader('Content-Type', 'text/event-stream');
        res.setHeader('Cache-Control', 'no-cache');
        res.setHeader('Connection', 'keep-alive');
        res.setHeader('Access-Control-Allow-Origin', '*');

        // Ganti .stream() dengan .create({ stream: true })
        const stream = await client.chat.completions.create({
        model: 'llama-3.3-70b-versatile',
        messages: [
            { role: 'system', content: systemPrompt },
            { role: 'user', content: userMessage }
        ],
        max_tokens: 1024,
        stream: true
        });

        for await (const chunk of stream) {
        const text = chunk.choices[0]?.delta?.content || '';
        if (text) res.write(`data: ${JSON.stringify({ text })}\n\n`);
        }

        res.write(`data: ${JSON.stringify({ done: true })}\n\n`);
        res.end();

    } catch (error) {
        console.error('Chatbot error:', error.message);
        res.write(`data: ${JSON.stringify({ error: 'Terjadi kesalahan.' })}\n\n`);
        res.end();
    }
}

module.exports = { streamChatResponse };

File: src/backend/ai/chatbot.py
# ai/chatbot.py
import os
from groq import Groq
from ai.prompts.systemPrompt import build_system_prompt

client = Groq(api_key=os.getenv("GROQ_API_KEY"))

def stream_chat_response(user_message: str, user_data: dict):
    """
    Generator function — yield chunk text satu per satu.
    FastAPI akan stream ini ke frontend.
    """
    system_prompt = build_system_prompt(user_data)

    stream = client.chat.completions.create(
        model="llama-3.3-70b-versatile",
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user",   "content": user_message}
        ],
        max_tokens=1024,
        stream=True
    )

    for chunk in stream:
        text = chunk.choices[0].delta.content or ""
        if text:
            yield text